thrimbletrimmer/scripts/stream.js

```diff
@@ -222,3 +228,94 @@ function convertEnteredTimes() {
 		document.getElementById("time-converter-from-ago").checked = true;
 	}
 }
+
+async function getStreamChatLog() {
+	const startTime = getStartTime();
+	const endTime = getEndTime();
+	if (!startTime || !endTime) {
+		return;
+	}
+	return getChatLog(wubloaderTimeFromDateTime(startTime), wubloaderTimeFromDateTime(endTime));
+}
+
+function initialChatRender() {
+	if (!globalChatData) {
+		return;
+	}
+	const videoPlayer = document.getElementById("video");
+	const videoTime = videoPlayer.currentTime;
+	const videoDateTime = dateTimeFromVideoPlayerTime(videoTime);
+	const chatReplayContainer = document.getElementById("chat-replay");
+	chatReplayContainer.innerHTML = "";
+
+	for (const chatMessage of globalChatData) {
+		if (chatMessage.when > videoDateTime) {
+			break;
+		}
+		handleChatMessage(chatReplayContainer, chatMessage);
+	}
+
+	globalChatPreviousRenderTime = videoTime;
+}
+
+function updateChatRender() {
+	if (!globalChatData) {
+		return;
+	}
+	const videoPlayer = document.getElementById("video");
+	const videoTime = videoPlayer.currentTime;
+
+	if (videoTime < globalChatPreviousRenderTime) {
+		initialChatRender();
+	} else {
+		const videoDateTime = dateTimeFromVideoPlayerTime(videoTime);
+		const lastAddedTime = dateTimeFromVideoPlayerTime(globalChatPreviousRenderTime);
+		const chatReplayContainer = document.getElementById("chat-replay");
+
+		let rangeMin = 0;
+		let rangeMax = globalChatData.length;
+		let lastChatIndex = Math.floor((rangeMin + rangeMax) / 2);
+		while (rangeMax - rangeMin > 1) {
+			if (globalChatData[lastChatIndex].when === lastAddedTime) {
+				break;
+			}
+			if (globalChatData[lastChatIndex].when < lastAddedTime) {
+				rangeMin = lastChatIndex;
+			} else {
+				rangeMax = lastChatIndex;
+			}
+			lastChatIndex = Math.floor((rangeMin + rangeMax) / 2);
+		}
```

Comment:
binary search, nice :ok_hand:

Reply:
yeah, I figured chat logs could be numerous (with DB chat being... DB chat, and particularly with long time ranges) and I wanted to be able to find the right chat quickly if possible.